arch/arm/core/aarch64/arm_mmu.c

```diff
@@ -340,15 +344,6 @@ static const struct arm_mmu_region mmu_zephyr_regions[] = {
 			      (uintptr_t)_image_rodata_start,
 			      (uintptr_t)_image_rodata_size,
 			      MT_RODATA | MT_SECURE),
-
-	/* Mark rest of the zephyr execution regions (data, bss, noinit, etc.)
-	 * cacheable, read-write
-	 * Note: read-write region is marked execute-ever internally
-	 */
-	MMU_REGION_FLAT_ENTRY("zephyr_data",
-			      (uintptr_t)__kernel_ram_start,
-			      (uintptr_t)__kernel_ram_size,
-			      MT_NORMAL | MT_RW | MT_SECURE),
```

Comment:
Basic idea here is to mark the **zephyr kernel regions**, such as text, rodata and data sections appropriately. And yes, these come from linker symbols. We do not want to mark the entire memory regions (such as SRAM) in here.

We removed SRAM marking from qemu_cortex_a53 as it was indicated that the entire SRAM is covered through regions marked here. If that is not the case anymore, please revert and put SRAM region back into qemu_cortex_a53 file rather than doing this.

Reply:
> Basic idea here is to mark the **zephyr kernel regions**, such as text, rodata and data sections appropriately. And yes, these come from linker symbols. We do not want to mark the entire memory regions (such as SRAM) in here.

We want to create an identity mapping for the whole SRAM and for that we do not have the linker symbols, so we are using the `CONFIG_*` symbols. We want to do this because this is a preliminary work to move to user-mode and user threads that are mapped outside the zephyr kernel regions.

> We removed SRAM marking from qemu_cortex_a53 as it was indicated that the entire SRAM is covered through regions marked here. If that is not the case anymore, please revert and put SRAM region back into qemu_cortex_a53 file rather than doing this.

Why? This is something that we always want, not something that should be SoC-specific.

Reply:
> > Basic idea here is to mark the **zephyr kernel regions**, such as text, rodata and data sections appropriately. And yes, these come from linker symbols. We do not want to mark the entire memory regions (such as SRAM) in here.
> 
> We want to create an identity mapping for the whole SRAM and for that we do not have the linker symbols, so we are using the `CONFIG_*` symbols. We want to do this because this is a preliminary work to move to user-mode and user threads that are mapped outside the zephyr kernel regions.
> 
No objection to mapping the whole SRAM or using the `CONFIG_*` symbols. I was just pointing out that purpose of marking zephyr kernel region is being satisfied by linker symbols.

> > We removed SRAM marking from qemu_cortex_a53 as it was indicated that the entire SRAM is covered through regions marked here. If that is not the case anymore, please revert and put SRAM region back into qemu_cortex_a53 file rather than doing this.
> 
> Why? This is something that we always want, not something that should be SoC-specific.

yes, I guess that's right, SRAM (CONFIG_SRAM_BASE) is always used for all SoCs per current linker script. However, data section may/may not be part of SRAM if custom linker script is used, so should not be removed.

Reply:
> yes, I guess that's right, SRAM (CONFIG_SRAM_BASE) is always used for all SoCs per current linker script. However, data section may/may not be part of SRAM if custom linker script is used, so should not be removed.

But if you have a custom linker script shouldn't then the mapping of the data section be done in the board / platform specific file?

Reply:
No.. marking of zephyr kernel regions is abstracted from user.

Reply:
That's not my point. If you have the `zephyr_data` region mapped outside the SRAM in some custom place (for example somewhere in DRAM) then that is something that you should specify in a MMU region defined in the board / soc file. In your case should be somewhere in `soc/arm/bcm_vk/viper/mmu_regions.c`. In the standard case the `zephyr_data` region is inside the SRAM with the same flags / permissions, I don't think why we should specify that as a standalone region.

Reply:
@sandeepbrcm I'm not exactly sure what you are suggesting here because the SRAM is defined as a separate entry in this patch.

Bottom line is that we need to have the SRAM region mapped and we cannot have a platform without `CONFIG_SRAM*` so this is something that we should have in the core driver. Otherwise we force every single platform to have a MMU region to map the SRAM, this is nonsense. FWIW this is following what's being already done by X86 (see https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/x86/gen_mmu.py).

@andrewboie @ioannisg any take  on this?

Reply:
I am suggesting to try
1- Add this entry for SRAM , because we know its an SRAM from the defines.
2- Retain   `zephyr_data `mapping using ` __kernel_ram_start`,  `__kernel_ram_size`  So that zephyr OS does not rely on any configs but map its own regions from linker symbols.

Reply:
But why considering that in the standard case`zephyr_data` region is a subset of the SRAM region with the same attributes / permissions?

Reply:
- map SRAM ( default attributes)
- map rest of the sections with their attributes.  zephyr_data section happens to be same as default

 It is only making this `zephyr_data`  section entry explicit in code, like a documentation without any assumption of SRAM map. Seems harmless . After all the final table entries are same either case.

Reply:
It depends on your definition of harmless. The way the code builds the page tables is not smart enough to detect that the two memory regions are overlapping so the whole `zephyr_data` section is mapped again in a different table and it even creates an additional table to host the duplicate entries.

Reply:
> the whole `zephyr_data` section is mapped again in a different table and it even creates an additional table to host the duplicate entries.

Let me correct you here. Additional table is not created for `zephyr_data` section. If SRAM has been entirely mapped and, `zephyr_data` section is mapped again; considering it is part of SRAM, logic will just do the mapping at the corresponding PTEs and PTEs will effectively remain unchanged as both SRAM and `zephyr_data` section have same attributes.

Now assume, if `zephyr_data` section is not in SRAM, then user does not need to worry about mapping, zephyr itself will take care of it, precisely why we are requesting to keep `zephyr_data` section entry explicit in code.

Reply:
> Let me correct you here. Additional table is not created for `zephyr_data` section. If SRAM has been entirely mapped and, `zephyr_data` section is mapped again; considering it is part of SRAM, logic will just do the mapping at the corresponding PTEs and PTEs will effectively remain unchanged as both SRAM and `zephyr_data` section have same attributes.

Are you sure about this? Because enabling the debugging it looks like this is not the case.

> Now assume, if `zephyr_data` section is not in SRAM, then user does not need to worry about mapping, zephyr itself will take care of it, precisely why we are requesting to keep `zephyr_data` section entry explicit in code.

I think we are bikeshedding again honestly. I'll add back the `zephyr_data` back so we can move to some more beefy PR about MMU ;)

Reply:
> > Let me correct you here. Additional table is not created for `zephyr_data` section. If SRAM has been entirely mapped and, `zephyr_data` section is mapped again; considering it is part of SRAM, logic will just do the mapping at the corresponding PTEs and PTEs will effectively remain unchanged as both SRAM and `zephyr_data` section have same attributes.
> 
> Are you sure about this? Because enabling the debugging it looks like this is not the case.
> 
yes, I am sure. Please check PTE address for both, which part of `MMU_DEBUG` prints (to see PTE address, please set  `DUMP_PTE` to 1).

> > Now assume, if `zephyr_data` section is not in SRAM, then user does not need to worry about mapping, zephyr itself will take care of it, precisely why we are requesting to keep `zephyr_data` section entry explicit in code.
> 
> I think we are bikeshedding again honestly. I'll add back the `zephyr_data` region so we can move to some more beefy PR about MMU ;)

important  to get basics right :)

Reply:
I can speak to the intent behind the x86 code.

What we produce on x86 is a build time set of page tables with all RAM addresses identity mapped. This is so these page tables can be installed at early boot and the kernel can continue execution as normal. On x86_64, you can't enter 64-bit mode without page tables.

What's really needed at minimum is to have all the RAM used by the kernel at boot, which in turn is defined by where things get linked. We need all program text, read-only data, and data mapped to where they are linked. If symbols are linked in a RAM region, we will need a mapping for it. The kernel assumes no runtime memory mappings are necessary for any RAM that is part of the linked kernel image once we get into z_cstart().

In addition, if `CONFIG_SRAM_REGION_PERMISSIONS` is turned on, then program text, read-only data, and data are in their own page-aligned regions and have read/execute/user, read/user, and read/write permissions respectively.

We might need some special handling if we have a physical memory map with sparse RAM regions. For example, the kernel assumes that virtual memory mappings may start at virtual address (CONFIG_SRAM_BASE_ADDRESS + KB(CONFIG_SRAM_SIZE)) up to (CONFIG_SRAM_BASE_ADDRESS + CONFIG_KERNEL_VM_SIZE). But if there is some other RAM region we need to idenity-map that overlaps this, we could get into trouble.

However, there is no requirement that RAM is identity-mapped. It just needs to all be mapped in the address space, and a defined region for additional memory mappings. The kernel will work fine if you link it virtually. So if you have multiple RAM regions, it may be better to just link the kernel at some fixed virtual address, and have all the RAM mappings, discontinuous physically, be one continuous region in virtual memory. For most systems identity-mapped RAM is just fine but you aren't forced to do it. The Kconfigs used by kernel/mmu.c may need some revision however.

If you don't identity map ram your arch code will be more complex as will be your linker script. I have a branch here on x86 where I was experimenting with it: https://github.com/andrewboie/zephyr/tree/vm-linking I also used a technique where the kernel was double-linked in the page tables at both its virtual and physical base addresses to transition the instruction pointer to the virtual domain at early boot. If this is of interest I can try to rebase this work.

x86 currently also identity maps *all* of system RAM in these page tables. This includes "unused" pages, past the end of the kernel image, to the RAM boundary. These "unused" pages are currently used to create the newlib heap in a way that doesn't involve memory mappings, but this is expected to change soon, along with other infrastructure for anonymous memory mappings and a core kernel ontology of physical page frames. But even with this, we'll probably still keep the identity mapping for these pages (they would be mapped twice, or re-mapped) so I don't expect this to change things much from your perspective.

HTH
Andrew

Reply:
the tl;dr version: if multiple physical RAM regions exist I think we need to consider not identity-mapping RAM and instead mapping it all in one virtual region, with linkage at these regions. I can dig up some prototyping I did for this on x86, I had this working well on 32-bit.

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/29803 should allow this on the core kernel side.